File: packages/nitro-renderer/src/room/object/RoomPlaneData.ts
import { type IVector3D, Vector3d } from '@nitrodevco/nitro-api';

import { RoomPlaneMaskData } from './RoomPlaneMaskData';

export class RoomPlaneData {
    public static PLANE_UNDEFINED: number = 0;
    public static PLANE_FLOOR: number = 1;
    public static PLANE_WALL: number = 2;
    public static PLANE_LANDSCAPE: number = 3;
    public static PLANE_BILLBOARD: number = 4;

    private _type: number = 0;
    private _loc: Vector3d = new Vector3d();
    private _leftSide: Vector3d = new Vector3d();
    private _rightSide: Vector3d = new Vector3d();
    private _normal: Vector3d = new Vector3d();
    private _normalDirection: Vector3d = new Vector3d();
    private _secondaryNormals: Vector3d[] = [];
    private _masks: RoomPlaneMaskData[] = [];

    constructor(type: number, loc: IVector3D, leftSideLoc: IVector3D, rightSideLoc: IVector3D, _arg_5: IVector3D[]) {
        this._type = type;
        this._loc.assign(loc);
        this._leftSide.assign(leftSideLoc);
        this._rightSide.assign(rightSideLoc);

        if (leftSideLoc && rightSideLoc) {
            this._normal = Vector3d.crossProduct(leftSideLoc, rightSideLoc);
            let normalX = 0;
            let normalY = 0;

            if (this.normal.x > 0 || this.normal.y > 0) {
                let x = this.normal.x;
                let y = this.normal.y;

                normalX = 360 + (Math.atan2(y, x) / Math.PI) * 180;

                if (normalX >= 360) normalX = normalX - 360;

                x = Math.sqrt(this.normal.x * this.normal.x + this.normal.y * this.normal.y);
                y = this.normal.z;

                normalY = 360 + (Math.atan2(y, x) / Math.PI) * 180;

                if (normalY >= 360) normalY = normalY - 360;
            } else if (this.normal.z < 0) {
                normalY = 90;
            } else {
                normalY = 270;
            }

            this._normalDirection = new Vector3d(normalX, normalY, 0);
        }

        if (!(_arg_5 == null) && _arg_5.length > 0) {
            let i = 0;

            while (i < _arg_5.length) {
                const _local_12 = _arg_5[i];

                if (_local_12 && _local_12.length > 0) {
                    const normal = new Vector3d();

                    normal.assign(_local_12);
                    normal.multiply(1 / normal.length);

                    this._secondaryNormals.push(normal);
                }

                i++;
            }
        }
    }

    public get type(): number {
        return this._type;
    }

    public get loc(): IVector3D {
        return this._loc;
    }

    public get leftSide(): IVector3D {
        return this._leftSide;
    }

    public get rightSide(): IVector3D {
        return this._rightSide;
    }

    public get normal(): IVector3D {
        return this._normal;
    }

    public get normalDirection(): IVector3D {
        return this._normalDirection;
    }

    public get secondaryNormalCount(): number {
        return this._secondaryNormals.length;
    }

    public get maskCount(): number {
        return this._masks.length;
    }

    public getSecondaryNormal(index: number): IVector3D {
        const loc = new Vector3d();

        if (index >= 0 && index < this.secondaryNormalCount) loc.assign(this._secondaryNormals[index]);

        return loc;
    }

    public addMask(leftSideLoc: number, rightSideLoc: number, leftSideLength: number, rightSideLength: number): void {
        this._masks.push(new RoomPlaneMaskData(leftSideLoc, rightSideLoc, leftSideLength, rightSideLength));
    }

    private getMask(k: number) {
        if (k < 0 || k >= this.maskCount) return undefined;

        return this._masks[k];
    }

    public getMaskLeftSideLoc(k: number) {
        return this.getMask(k)?.leftSideLoc ?? -1;
    }

    public getMaskRightSideLoc(k: number) {
        return this.getMask(k)?.rightSideLoc ?? -1;
    }

    public getMaskLeftSideLength(k: number): number {
        return this.getMask(k)?.leftSideLength ?? -1;
    }

    public getMaskRightSideLength(k: number): number {
        return this.getMask(k)?.rightSideLength ?? -1;
    }
}

File: packages/nitro-swift/Sources/NitroRoom/Plane/RoomPlaneData.swift
import Foundation

import NitroCore

/// One rectangle-shaped mask baked directly onto a plane at parse time (distinct from the dynamic,
/// server-driven bitmap mask system in `PlaneMaskManager`). Mirrors `RoomPlaneMaskData`.
public struct RoomPlaneMaskData {
    public let leftSideLoc: Double
    public let rightSideLoc: Double
    public let leftSideLength: Double
    public let rightSideLength: Double
}

/// Swift port of `RoomPlaneData` (packages/nitro-renderer/src/room/object/RoomPlaneData.ts) - the
/// geometry record for one floor/wall/landscape polygon (a parallelogram: `loc` is one corner,
/// `leftSide`/`rightSide` are edge vectors).
public final class RoomPlaneData {
    public static let planeUndefined = 0
    public static let planeFloor = 1
    public static let planeWall = 2
    public static let planeLandscape = 3
    public static let planeBillboard = 4

    public let type: Int
    public let loc: Vector3d
    public let leftSide: Vector3d
    public let rightSide: Vector3d
    public let normal: Vector3d
    /// `x` = yaw (0-360), `y` = pitch (0-360) of the plane's facing direction - used to bucket
    /// which bitmap-mask art variant applies to this plane (see `PlaneMaskVisualization`).
    public let normalDirection: Vector3d
    public private(set) var secondaryNormals: [Vector3d] = []
    public private(set) var masks: [RoomPlaneMaskData] = []

    public init(type: Int, loc: Vector3d, leftSide: Vector3d, rightSide: Vector3d, secondaryNormals: [Vector3d]) {
        self.type = type
        self.loc = Vector3d.from(loc)
        self.leftSide = Vector3d.from(leftSide)
        self.rightSide = Vector3d.from(rightSide)

        let normal = Vector3d.crossProduct(leftSide, rightSide)

        self.normal = normal

        var normalX: Double = 0
        var normalY: Double = 0

        if normal.x > 0 || normal.y > 0 {
            var x = normal.x
            var y = normal.y

            normalX = 360 + (atan2(y, x) / .pi) * 180
            if normalX >= 360 { normalX -= 360 }

            x = (normal.x * normal.x + normal.y * normal.y).squareRoot()
            y = normal.z

            normalY = 360 + (atan2(y, x) / .pi) * 180
            if normalY >= 360 { normalY -= 360 }
        } else if normal.z < 0 {
            normalY = 90
        } else {
            normalY = 270
        }

        normalDirection = Vector3d(normalX, normalY, 0)

        for secondaryNormal in secondaryNormals where secondaryNormal.length > 0 {
            let normalized = Vector3d.from(secondaryNormal)

            normalized.multiply(1 / normalized.length)
            self.secondaryNormals.append(normalized)
        }
    }

    public func getSecondaryNormal(_ index: Int) -> Vector3d {
        guard index >= 0, index < secondaryNormals.count else { return Vector3d() }

        return Vector3d.from(secondaryNormals[index])
    }

    public func addMask(leftSideLoc: Double, rightSideLoc: Double, leftSideLength: Double, rightSideLength: Double) {
        masks.append(RoomPlaneMaskData(leftSideLoc: leftSideLoc, rightSideLoc: rightSideLoc, leftSideLength: leftSideLength, rightSideLength: rightSideLength))
    }
}
